Dear user,<br>
<p>
We are happy to announce that we have released <a href="http://cab.spbu.ru/software/spades/">SPAdes 3.14</a>. This release includes such new features as:<br>
- BiosyntheticSPAdes pipeline for identifying Biosynthetic Gene Clusters (BGCs);<br>
- Hybrid transcriptome assembly (Illumina RNA-Seq + Iso-seq / ONT RNA);<br>
- Plasmid assembly from metagenomic datasets;<br>
- New --isolate option that improves assembly quality of standard datasets with high coverage;<br>
- New standalone tools for read filtration based on k-mer coverage and estimating number of unique k-mers in reads;<br>
- Improved SPAligner tool for read-to-graph alignment;<br>
- Major performance improvements and memory consumption reduction in graph construction,simplification procedures and throughout the whole pipeline;<br>
- Several bug-fixes.<br>
Please, consider updating.
<p>
SPAdes 3.14.0 is released under GPLv2 and openly available for download from our website <a href="http://cab.spbu.ru/software/spades/">cab.spbu.ru/software/spades/</a>
<p>
Your comments, bug reports, and suggestions are very welcomed. They will help us to further improve SPAdes. You can look for existing issue on our <a href="https://github.com/ablab/spades">GitHub repository</a>, create a new one or write us via e-mail:  <a href="mailto:spades.support@cab.spbu.ru">spades.support@cab.spbu.ru</a>. If you have any troubles running SPAdes, please provide us with the files params.txt and spades.log from the output directory.

<p>
Best regards, <br>
SPAdes team

