New SPAdes 3.7
<p>
Dear user, <br>
<p>
We are happy to announce that we have just released new <a href="http://bioinf.spbau.ru/en/spades">SPAdes 3.7.0</a>. This version includes the following modifications: <br>
<p>
- New truSPAdes â€“ an assembler for short reads produced by Illumina TruSeq Long Read technology; <br>
- Better running time, less RAM consumption and improved results for BayesHammer error correction module;<br>
- Improvements and bugfixes in dipSPAdes;<br>
- Multiple improvements and bugfixes in repeat resolution and scaffolding modules;<br>
- MismatchCorrector module now uses bwa-mem; several bugfixes are made.<br>
<p>
Please note that SPAdes was designed for single-cell and standard bacterial and other small (e.g. fungal) data sets and was not intended for larger genomes (e.g. mammalian size genomes). Should you decide to use SPAdes for a larger genome, please be aware that you are doing it at your own risk.
<p>
SPAdes version 3.7.0 is released under GPLv2 and openly available for download from <a href="http://bioinf.spbau.ru/en/spades">http://bioinf.spbau.ru/en/spades</a>.
<p>
Your comments, bug reports, and suggestions are very welcomed. They will help us to further improve SPAdes. If you have any troubles running SPAdes, please provide us with the files <b>params.txt</b> and <b>spades.log</b> from the output directory. Address for communications: <a href="mailto:spades.support@bioinf.spbau.ru">spades.support@bioinf.spbau.ru</a>.<br>
<p>
Best regards, <br>
SPAdes team
