configfile: "config.yaml"
GAPATH = config["spalignerpath"]
GACFG = config["spalignercfg"]
GRAPH = config["graph"] 
K = config["kmer"]

ALL_ORGS=["ecoli", "celegans", "scerevisiae"]
ALL_READS=["simpb2000", "simnp2000","realpb2000", "realnp2000"]

SCRIPT_PATH=os.getcwd()

workdir: config["datapath"]

rule all:
    input:
        expand("{org}/SPAligner/output/aln_{reads}.tsv", reads=ALL_READS, org=ALL_ORGS)


rule run_aligner:
    params:
        tp = lambda wildcards: "pacbio" if wildcards.reads.endswith("pb2000") else "nanopore"
    output:
        tsv = "{org}/SPAligner/output/aln_{reads}.tsv"
    log:
        "{org}/SPAligner/output/aln_{reads}.log"
    benchmark:
        repeat("{org}/SPAligner/benchmark/align_{reads}.tsv", 1)
    threads: 16
    run:
        shell("{GAPATH}/longreads_aligner {SCRIPT_PATH}/{GACFG} -g {wildcards.org}/{GRAPH} \
                -s {wildcards.org}/input/{wildcards.reads}.fasta -K {K} \
                -d {params.tp} -o {wildcards.org}/SPAligner/output/aln_{wildcards.reads} > {log}")